PIC runtime for 32-bit ARM with __PIC__

jonahbeckford opened the pull request:
Since Android requires position independent code with the security feature [Text Relocations (Enforced for API level >= 23)](https://android.googlesource.com/platform/bionic/+/master/android-changes-for-ndk-developers.md#text-relocations-enforced-for-api-level-23) | ([permalink](https://android.googlesource.com/platform/bionic/+/e2c784f15912115f778e427f0ce9a21189a3b494/android-changes-for-ndk-developers.md#text-relocations-enforced-for-api-level-23)) the existing arm32 native code fails to run on Android when using API 23+.

Android does officially support [32bit ARM: armeabi-v7a](https://developer.android.com/ndk/guides/abis#sa). You'll get an "Illegal Instruction" running native OCaml code (ex. in the Android Emulator) in the following absolute relocations of `arm.S`:

```assembly
        ldr     r12, =caml_program

        ldr     domain_state_ptr, =Caml_state

        ldr     r5, =.Ltrap_handler

        ldr     r12, =caml_apply2

        ldr     r12, =caml_apply3

        ldr     r7, =caml_array_bound_error
````

To make PIC code both 1) absolute relocations have to be removed and 2) the indirection of PLT needs to be used for external functions.

This PR does that in `arm.S` if and only if `__PIC__` is defined.

The [original commit Merge the new ARM backend into trunk (PR#5433)](https://github.com/ocaml/ocaml/commit/05627e0de4a50067f36d1eca9dcc9ebd5736c3f8) that introduced PIC code for ARM did not fully complete the PIC transition (aka. make a 100% position independent executable) because although the OCaml compiler emitted code was PIC compliant, it still used absolute relocations in the glue code `arm.S`. For example https://github.com/ocaml/ocaml/blame/cb9826b1322a64a6f760de287848b73c7e429ee5/asmrun/arm.S#L161:

```assembly
        ldr     r12, =caml_young_limit
```

---

Tested was a bit complicated; I used a [4.12.1 cross-compilation apparatus](https://gitlab.com/diskuv/diskuv-ocaml/-/blob/next/installtime/unix/private/reproducible-compile-ocaml-1-setup.sh) to cross an Ubuntu Linux build machine with an 32-bit ARM Android Emulator host machine. I also used lldb + lldb-server to step through the runtime code (arm.S) and generated code for verification. At the end I was able to build and run:

```ocaml
let () = print_endline "Hello World!"
```

with `ocamlopt.opt hello_world.ml -o hello_world.armeabi-v7a.opt.exe`.

> I suspect this PR will be sitting for a while waiting for the Multicore OCaml code freeze. When this passes a sniff test from a core dev I'll update the cross-compiler apparatus from 4.12.1 to whatever the latest OCaml is, and then do another retest.

The basic OCaml configuration was Android NDK 23 toolchain with NDK's GNU AS for the assembler and NDK's clang compiler for C. All flags you see below come from the toolchain (I added `-g3` and `-fdebug-macro` for debugging, and `-Wl,--fix-cortex-a8` since it was recommended).

```sh
AR="$HOME/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/llvm-ar" \
AS="$HOME/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/arm-linux-androideabi-as -g" \
ASPP="$HOME/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/clang --target=armv7-none-linux-androideabi21  -fPIE -fPIC -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -march=armv7-a -mthumb -mfpu=vfpv3-d16 -Wformat -Werror=format-security -fexceptions -g3 -g  -fno-limit-debug-info -c -fdebug-macro" \
CC="$HOME/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/clang --target=armv7-none-linux-androideabi21 --sysroot=$HOME/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/sysroot" \
CFLAGS=" -fPIE -fPIC -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -march=armv7-a -mthumb -mfpu=vfpv3-d16 -Wformat -Werror=format-security -fexceptions -g3 -g  -fno-limit-debug-info" \
CPP="$HOME/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/clang --target=armv7-none-linux-androideabi21 --sysroot=$HOME/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/sysroot -E" \
DIRECT_LD="$HOME/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld" \
LD="$HOME/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld" \
LDFLAGS=" -fPIE -pie -Wl,--fix-cortex-a8" \
NM="$HOME/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/llvm-nm" \
OBJDUMP="$HOME/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/llvm-objdump" \
RANLIB="$HOME/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/llvm-ranlib" \
RANLIBCMD="$HOME/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/llvm-ranlib" \
STRIP="$HOME/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/llvm-strip" \
./configure \
    --prefix "/tmp/$(git rev-parse --abbrev-ref HEAD)" \
    --with-sysroot="$HOME/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/sysroot" \
    --host=armv7-none-linux-androideabi21 \
    --disable-ocamldoc
```

jonahbeckford committed:
PIC runtime for arm32 with __PIC__

./configure and make with CC=-D__PIC__ or -fPIC for compilers that automatically set __PIC__
will create position independent code for asm.o and libasmrun.a.

The code path when __PIC__ is not defined remains untouched. Defining __PIC__ will increase
the code size of the ARM runtime libasmrun.a slightly, add three (3) frame descriptors,
and make use of PLT indirection for external function linkage.

Renamed Caml_state() macro to Caml_state_var() macro to unshadow compiler expressions, especially
Caml_state(GOT_PREL)

nojb reviewed (commented):
I haven't looked in detail at the proposed change, but the PR description sounds reasonable. However, two general observations:

- The renaming `Caml_state => Caml_state_var` seems purely cosmetic. Normally we try to avoid such renamings because they add noise to the diff and to the history (if they are committed). **If** there is a genuine reason for this renaming it should be adressed on a dedicated PR and it should also adapt all the other backends at the same time.

- The off-by-one fix in `roots_nat.c` seems a genuine fix that should be looked at independently of this PR. Note that this code is being reworked in the Multicore PR #10831; it is worth checking the corresponding code in that PR and fix it there directly.

xavierleroy commented:
Mentioning @EduardoRFS, who is an expert on OCaml on Android.

Globally this looks nice, and I'm relieved no changes are needed  to the code generator (asmcomp/arm/emit.mlp), all changes are in the glue code between the generated code and the runtime system.

Is it the case that ARM+PIC used to work in the past and was broken in a recent release?  e.g. following the introduction of the Caml_state structure?  Just curious.  Also, if it's the case, it could help reviewing this PR.

jonahbeckford commented:
> * @nojb The renaming `Caml_state => Caml_state_var` seems purely cosmetic. Normally we try to avoid such renamings because they add noise to the diff and to the history (if they are committed). **If** there is a genuine reason for this renaming it should be adressed on a dedicated PR and it should also adapt all the other backends at the same time.

It is not cosmetic. Adding the following to the commit message and to "Changes": `Renamed Caml_state() macro to Caml_state_var() macro to unshadow compiler expressions, especially Caml_state(GOT_PREL).`

And propagating the same renaming to the other runtime assembly glue (i386.S, etc.).

> * @nojb The off-by-one fix in `roots_nat.c` seems a genuine fix that should be looked at independently of this PR. Note that this code is being reworked in the Multicore PR #10831; it is worth checking the corresponding code in that PR and fix it there directly.

Since the merge already happened (congrats!) I peeled it off into a separate PR branched off the post-merge trunk.

nojb commented:
> It is not cosmetic.

Right, I had missed that in your patch, sorry.

jonahbeckford commented:
> * @xavierleroy Is it the case that ARM+PIC used to work in the past and was broken in a recent release? e.g. following the introduction of the Caml_state structure? Just curious. Also, if it's the case, it could help reviewing this PR.

I'll update the PR description with the following ...

The non-PIC code are the locations below that use absolute relocations:

```assembly
        ldr     r12, =caml_program

        ldr     domain_state_ptr, =Caml_state

        ldr     r5, =.Ltrap_handler

        ldr     r12, =caml_apply2

        ldr     r12, =caml_apply3

        ldr     r7, =caml_array_bound_error
````

To make PIC code both 1) absolute relocations have to be removed and 2) the indirection of PLT needs to be used for external functions.

This PR does that for arm.S.

The [original commit Merge the new ARM backend into trunk (PR#5433)](https://github.com/ocaml/ocaml/commit/05627e0de4a50067f36d1eca9dcc9ebd5736c3f8) that introduced PIC code for ARM did not fully complete the PIC transition (aka. make a 100% position independent executable) because although the OCaml compiler emitted code was PIC compliant, it still used absolute relocations in the glue code arm.S.

Note: If we have a CI test apparatus somewhere I (or someone) can update it with a `readelf` test to check for the absence of absolute relocations in a test OCaml generated executable.

jonahbeckford committed:
Propagate Caml_state_var() rename

Some side comments needed trimming as well to meet the 80-col code hygiene standard.

EduardoRFS commented:
Will be looking on this PR soon, I hadn't noticed the runtime for ARM is not position independent, I think it's because my default API target is Android 21.

edit: will be looking at it on 21/01, but I find this patch very weird, especially as armv7-a on clang doesn't generate PLT
